### ISSUE TITLE
Intent training disturbs entity training? Bug?

### ISSUE THREAD

**Issue by ctrado18** [RasaHQ/rasa_nlu#1448]
I thought I could train NER_CRF and intent classification (tensorflow) independently?

What I do:

1. NER_CRF

```
{"rasa_nlu_data":{"regex_features":[],"entity_synonyms":[],"lookup_tables": [
            {
                "name": "entity",
                "elements": "words.txt"
            }
        ],
"common_examples":[{
"text":"Hr bla bala",
"intent":"",
"entities":[{"end":13,"entity":"entity","start":8,"value":"bla"}]}

```

I train it with setting `intent:""` such that no intent is trained. 

2. Train intent: This contains blocks where entities are removed and just labeled with intent

```
{
"text":"bal bla bla",
"intent":"MyIntent",
}
```

Then I merged both parts together. But now I realise that my good NER is destroyed including the intent training part?! How can this be since the `entities` field is removed?

My pipeline:

```

language: "de_core_news_sm"

pipeline:
- name: "intent_featurizer_count_vectors"
- name: "intent_classifier_tensorflow_embedding"
- name: "tokenizer_whitespace"
- name: "intent_entity_featurizer_regex"
- name: "ner_crf"
  "features": [
            ["prefix5", "prefix2", "suffix3",
             "suffix2", "title", "upper"],
            ["bias", "upper", "title", "digit", "pattern"],
            ["prefix5", "prefix2", "suffix3",
             "suffix2", "title", "upper"]]

```

For both parts and just the entity part the number of used entity examples re the same as seen from debug. That is really bad since I tried a lot to find good examples to train NER_CRF. I was so happy but no... It seems like a bug?

v0.13.3

I suppose a conflict with the regex featurizer as it is trained on all examples in train method? Although I just want to use it for entities. How can I manage that?

**Comment by akelad** [comment]
What do you mean by "destroyed"? 
How come you're training these separately though? Also, did you train this model on two different datasets using the same pipeline?

**Comment by ctrado18** [comment]
@akelad 

destroyed : NER is bad.

I trained NER_CRF and Intent separately with different examples and merged them in above way (correct?!) in asszumption doing right?

And I think focus is on new lookup table as I hacked it.

I found the issue is beeing the regex featurizer! Ther, it take all training examples for learning features. Before lookup table just learns on my NER dfata set. This disturbes I think. I checked it right now  to change it to:

```
for example in training_data.entity_examples:
            example.set("text_features", updated)

```
Such only entity examples are taken.

This gives back my good NER as before!

But I think this is not good as maybe later I want to use Regex for intent or doing something else. ALthough I found that then you have to call regex after tensorflow embedding since it gives an error.

**Comment by akelad** [comment]
i'm still not sure what you mean by "NER is bad", what kind of performance change are we talking about? Also we don't recommend training these things separately, setting the intent to `""` doesn't actually mean it has no intent, it just means you're training an empty intent. 

And the regex ner has always been trained on all examples. this is because there might be certain cases where you don't want to extract these entities and therefore we need to incorporate the ones that aren't labeled as entities

**Comment by ctrado18** [comment]
> i'm still not sure what you mean by "NER is bad", what kind of performance change are we talking about? 

way more bad testing on same data. But you should focuse on the other parts and below. The point is why is there a change? There is still a change when just using the entity examples for regex but erformance is nearly as good as before. I don't get same performance as without the intent examples. 

> Also we don't recommend training these things separately, setting the intent to "" doesn't actually mean it has no intent, it just means you're training an empty intent.

It was recommended here in the repo. And you need to do this actually or how would you else train NER without doing overfitting. since you have to take care for different things training intent and NER:  I tested it and when using `` as intent it was not given as intent?
When removing entity dict from training data, this example is not trained with NER_CRF right?

> This is because there might be certain cases where you don't want to extract these entities and therefore we need to incorporate the ones that aren't labeled as entities

I don't undertsand this. But when you want to use regex for entity training than you should also taken those examples with entities?

**Comment by akelad** [comment]
Honestly it's hard to tell you anything about why your performance is "bad" (which i still don't have a metric for), without seeing what's in your regex list and your training data. 
Where was this recommended in the repo?
As said, I would recommend training entity extraction on *all* examples. If you don't, you might get a lot of false positives recognised.
I'd also like to move this discussion to the [forum](https://forum.rasa.com) now, so if you'd like to discuss this further please create a post there

**Comment by ctrado18** [comment]
Hey, thanks.

Before I open it in the forum, it would be nice if you comment on my phrases above. As I said don't focus on performance. It is "bad" and I thought this could be an issue for this repo. Since I like to use regex just for entities and not my intent examples. After shrinking it to entity examples it was resolved somehow the performance. I think it was @amn41 few times which states that with empty intent. So shrinking it to entities in regex is just for usage for the look up table! Since and here I would like to here also your comment, if the entities array is missing in examples they are not used to training right? Also, maybe i is better to remove for entity training the intent array at all instead of making it empty? Since no intent array the tensorflow does not use it?

And, it is a suitable way of handling to separate intent and entity training since both belong to different attentions.


Please comment on this . I feel most of the time that you don't pay much attention to the real issues ;-)

**Comment by ctrado18** [comment]
> no i think leaving out the entities field doesn't exclude that from training, you're just saying there's no entities present in the sentence, which the CRF still uses to train though.

That is really confusing as I read the code that only examples with `entities` field are used. Could you explain this more? I wonder how CRF can be trained withhout features for this example?

When setting `intent: ""` is tensoflow embedding satill uses this examples for training?! When looking in the model folder where also training data is copied you see that for those examples no intent field is there at all! SO the natural thinking is that it will not trained...

I encounter an issue right now. I use examples both for intent and entity training removing for each the specific label. They are the same examples. Now, for this one intent where only the word `problem` occurs this gets now classified as `None`. I feel that this is because of this word appears in examples with no intent label and in with the intent label?


> For all I know you could have words like "and" in there which is causing some terrible misclassification

I extra cheked this.


> One question I have: do you have in your intent examples some examples of sentences with entities that are in your lookup tables, but aren't labeled there?

Yes, but I shrinked the regex examples to just the `entitytrainingexamples`, so this should not matter for look up table?
